---
title: "PJ_ANGSD----Next Generation Sequencing Facilitates Quantitative Analysis of A549 Control and A549 infected with Zika virus Transcriptomes"
author: "JingjingQi"
date: "2/13/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(data.table)
library(DESeq2)
#BiocManager::install("org.Hs.eg.db")
library("TxDb.Hsapiens.UCSC.hg38.knownGene")
library(goseq)
library(org.Hs.eg.db)
```

```{css, echo=FALSE}
.scroll {
  max-height: 200px;
  overflow-y: auto;
}
```

### Introduction  

Zika virus (ZIKV) is an arthropod-borne virus(transmited by Arthropod) of the genus Flavivirus(with + sense ssRNA,  10–11kb in length). The recent ZIKV pandemic in theWestern Hemisphere was associatedwith a dramatic increase inmicrocephaly and other neurological deficits in the fetuses. In-vitro study on human fetal neural progenitors (hNPs) had shown that ZIKV replication persists in hNPs for at least 28 days.[1] hNPs can be differentiated into neurons, oligodendrocytes, or astrocytes. Human fetal astrocytes(HFAs) is the most abundant cell type in the central nervous system, they may function as composing the BBB, provision of nutrients to the neurons, maintenance of extracellular ion balance, and damage repair.[2] The research of this paper hypothesis that HFAs are reservoirs for ZIKV in the fetal brain that may contribute to the establishment of chronic brain infection and neurodevelopmental abnormalities.    

###Evidence:  
HFAs permissive to ZIKV infection.    
Low host cell apoptosis; Virus can thrive in the presence of a sustained cytokine antiviral response.    
qrtPCR ZIKV genomes in persistently infected HFAs suggest that adaptive mutations were not required for establishing chronic infection.   
RNAseq analyses of persistently infected cells revealed that ZIKV alters host gene expression in a manner that could affect developmental processes. *---------source of data is part of the main study ZIKV infected A549(Positive Control continuous human cell lines) vs non-infecte A549*    

### Method  

###Samples 
HFAs were obtained from 15 to 19-week aborted fetuses with written consent.    
Positive Control: continuous human cell lines A549 cells overexpress AXL with high permissive property.       
(HFAs were highly permissive to ZIKV infection, it is shown to be TIM/TAM receptor member AXL dependent.[3], A549 overexpress AXL, depending upon the donor, only 23–46% of HFAs exhibited detectable levels of this protein)   

###RNA Sequencing  
RNA of A549 cells-infected with Zika virus and un-infected A549 cells was isolated using RNeasy mini kit (QIAGEN, Valencia, CA, USA). RNAseq libraries were constructed using a TruSeq RNA Sample Prep V2 Kit. mRNA profiles was generated by deep sequencing, in triplicate, using Illumina Hiseq4000. All samples passed QC without triming.  

###Sequence Alignment and Gene Level Expression Quantification  
Sequence alignment and quantification of gene and exon level expression was carried out using RNA-seq analytical pipeline see _appendex_. Pair end reads were aligned to the human genome build hg38.refGene using star(2.7.0e), Star alignReads command options: --alignIntronMin 20 --alignIntronMax 1000000 --twopassMode Basic. Subread(1.6.2e) was used to count the reads mapping to individual exons according to hg38.refGene annotations (10-Jan-2020). Gene level expression was normalized by DESeq2(3.10e) Bioconductor packages, using Relative Log Expression (RLE) assuming most genes are not DE, that is calculated as the median of the ratio for each gene of its read counts over its geometric mean across all samples account for sequencing depth and RNA composition of the sample.  


### Result  

_table1._ Summary of average aligned reads count and average aligned reads percentage     
```{r, class.output="scroll", echo=F}
r_exon <- fread("data/featCounts_transcript_all.txt.summary.txt")
#rename <- gsub(".Aligned.sortedByCoord.out.bam$", "", x = colnames(r_exon))
#rename <- gsub("/athena/angsd/scratch/jiq4001/Pj1/Star_align/", "", x = rename)
#colnames(r_exon) <- rename
colnames(r_exon) <- c("Status", outer(c("Ctrl", "Infected"), c("Rep1", "Rep2", "Rep3"), paste))

r_exon %>%
  filter(Status %in% c("Assigned", "Unassigned_MultiMapping",
                       "Unassigned_NoFeatures", "Unassigned_Overlapping_Length",
                       "Unassigned_Ambiguity")) %>%
  gather(-Status, key = "sample", value = "reads") %>%
  group_by(sample) %>%
  mutate(percent = round(reads/sum(reads)*100, 2)) -> r_exon_meata

r_exon_meata %>%
  separate(sample, into = c("condition", "rep"))%>%
  group_by(Status) %>%
  summarise(average_aligned_reads_count = mean(reads)%>%round(0),
            average_aligned_percentage = mean(percent)%>%round(2)) %>%
  kableExtra::kable()%>%
  kableExtra::kable_styling()
```

The average aligned reads count and average aligned reads percentage of the 6 samples are summarized in _Table.1_. On average, 60.54% aligned reads are uniquely assigned, 26.21% aligned reads are with multi-mapping. 5.02% aligned reads are unassigned due to ambiguity, and less than 9% aligned reads are with no features or too short overlapping length. Individual reads count and percentage summary per sample is shown in the _Figure.1_.     

```{r, echo=F, warning=F, message=F}
#### bar graph alternatime
r_exon_meata %>%
  ggplot()+
  geom_bar(aes(sample, percent, fill = Status), stat = "identity", position = "stack")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        panel.background = element_rect(fill = "white"),
        axis.line = element_line(colour = "black")) -> p1

r_exon_meata %>%
  ggplot()+
  geom_bar(aes(sample, reads, fill = Status), stat = "identity", position = "stack")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        panel.background = element_rect(fill = "white"),
        axis.line = element_line(colour = "black"))-> p2


ggpubr::ggarrange(p1, p2, common.legend = T, legend = "bottom")
```


```{r, class.output="scroll", echo=F, message=F, warning=F}
# read data table
rc0 <- read.table("data/featCounts_transcript_all.txt", header=TRUE, row.names = NULL)


# trim extra text from colname
names(rc0) <- gsub("X.athena.angsd.scratch.jiq4001.Pj1.Star_align.", "", names(rc0))
names(rc0) <- gsub(".Aligned.sortedByCoord.out.bam", "", names(rc0))

# generate rowname with GeneId
row.names(rc0) <- make.names(rc0$Geneid, unique = T) 
readcounts <- rc0[ , -c(1:6)]


all <- readcounts
namedf <- outer(c("Ctrl", "Infected"), c("Rep1", "Rep2", "Rep3"), paste)
names(all) <- c(namedf[1,], namedf[2,])
all_info <- DataFrame(condition = rep(c("Ctrl", "Infected"), each = 3),
                      row.names = names(all))


# wrap reads, condation into object

DESeq.all <- DESeqDataSetFromMatrix(countData = all, colData = all_info, design = ~ condition)

# normalization
DESeq.all <- estimateSizeFactors(DESeq.all)
log.norm.counts <- log2(counts(DESeq.all, normalized=TRUE) + 1)
assay(DESeq.all, "log.norm.counts") <- log.norm.counts

```
`
Reads counts are normalized on per gene/sample bases with Deseq2 R package, followed by diffrencial expression analysis using GLM with NB distribution model.   

To investigate the differencially expressed genes, multiple testing on normalized reads count at 0.05 significent level with each gene using glm with contrast model ~ condition(condition_Infected_vs_Ctrl).   

```{r, echo=F, message=F, warning=F}
DESeq.all <- DESeq(DESeq.all)
#resultsNames(DESeq.all)
DESeq.all.result <- results(DESeq.all, independentFiltering = TRUE, alpha = 0.05,
                             name = "condition_Infected_vs_Ctrl")

#plotDispEsts(DESeq.all)
```

6258, 29% genes are excluded from multiple testing. the mean read counts(normalized) of 1265 out of 21856 genes are different in the infected group than control gorup with adjusted p-value < 0.05. Heatmap of the mean normalized Counts is plotted for genes with adjusted p-value < 0.05 in _Figure.2_.     

```{r, echo=F, message=F, warning=F, eval=F}
summary(DESeq.all.result)
table(DESeq.all.result$padj < 0.05)
hist(DESeq.all.result$padj)
plotMA(DESeq.all.result, alpha = 0.05, main = "Test: p.adj.value < 0.05", ylim = c(-8,8))
```
_Figure.2_. Heatmap of mean of Normalized Counts is plotted for genes with adjusted p-value < 0.05.      
```{r, echo=F, message=F}
DESeq.all.result.sorted <- DESeq.all.result[order(DESeq.all.result$padj), ]%>% head(table(DESeq.all.result$padj < 0.05)[2])

DE_gene <- DESeq.all@assays@data$log.norm.counts[rownames(DESeq.all.result.sorted),]
rowname_df <- data.frame(Condition = factor(rep(c("Ctrl", "Infected"), each = 3)), 
                         BioRep = factor(rep(c("Rep1", "Rep2", "Rep3"), 2)))
rownames(rowname_df) <- colnames(DE_gene)
pheatmap::pheatmap(DE_gene, scale = "row", annotation_col = rowname_df, show_rownames = F)

```
    
_Figure.3_ showed the Volcano plot of the Log2 fold change vs -Log10 adjusted P-value is plotted for DEgenes with adjusted p-value < 0.05. Genes with more than 2 fold log change is colored in red. 56 DE genes showed at least more than log2 fold change, among which 54 are positive. Most of differencial expressed genes with more than log2 fold change in the infected condition are upregulated compare to control non-infected condition.        
      
_Figure.3_ Volcano plot DEgenes with adjusted p-value < 0.05   
```{r, echo=F, warning=F, message=F}
### plot of DE genes aj-pvalue < 0.05
plot(y = -log10(DESeq.all.result.sorted$padj), 
     x = DESeq.all.result.sorted$log2FoldChange, 
     col = ifelse(abs(DESeq.all.result.sorted$log2FoldChange) >= 2, "red", "black"),
     xlab = "log2 fold change", ylab = "-log10 adjusted-p-value")
abline(v = c(-2, 2), col = "red", lty = 2)


DESeq.all.result.sorted.gt2fc <-      DESeq.all.result.sorted[which(abs(DESeq.all.result.sorted$log2FoldChange) >= 2), ] %>%data.frame()
```

 

```{r, echo=F, message=F, warning=F, eval=F}
gene.vector <- ifelse(DESeq.all.result$padj < 0.05 & abs(DESeq.all.result$log2FoldChange) >= 2, 1, 0)
#sum(gene.vector, na.rm = T)
names(gene.vector) <- rownames(DESeq.all.result)
gene.vector <- ifelse(is.na(gene.vector), 0, gene.vector)

#glen <- rc0$Length
glen <- getlength(rownames(DESeq.all.result), "hg38","geneSymbol")
#sum(gene.vector)
#length(glen)
#length(gene.vector)
gid <- rownames(rc0)
pwf <- nullp(gene.vector, id = gid, bias.data = glen)
go.wall <- goseq(pwf,"hg38","geneSymbol", use_genes_without_cat=T)
#go.gns <- getgo(rownames(DESeq.all.result), "hg38", "geneSymbol")%>%stack()
idx <- p.adjust(go.wall$over_represented_pvalue,method="BH")<.05 | p.adjust(go.wall$under_represented_pvalue,method="BH")<.05
go.wall <- go.wall[idx,]
#write_csv(go.wall, "go.wall_1.csv")


```
_Figure.4_ REVIGO Gene Ontology treemap of DEgene with at least log2 fold changes.        
```{r, echo=F, warning=F, message=F}
source("revigo_r/REVIGO_treemap.r")  
REVIGO_treemap <- function(revigo.data, col_palette = "Paired",
                           title = "REVIGO Gene Ontology treemap", ...){
  stuff <- data.frame(revigo.data)
  names(stuff) <- c("term_ID","description","freqInDbPercent","abslog10pvalue",
                    "uniqueness","dispensability","representative")
  stuff$abslog10pvalue <- as.numeric( as.character(stuff$abslog10pvalue) )
  stuff$freqInDbPercent <- as.numeric( as.character(stuff$freqInDbPercent) )
  stuff$uniqueness <- as.numeric( as.character(stuff$uniqueness) )
  stuff$dispensability <- as.numeric( as.character(stuff$dispensability) )
  # check the treemap command documentation for all possible parameters - 
  # there are a lot more
  treemap::treemap(
    stuff,
    index = c("representative","description"),
    vSize = "abslog10pvalue",
    type = "categorical",
    vColor = "representative",
    title = title,
    inflate.labels = FALSE,      
    lowerbound.cex.labels = 0,   
    bg.labels = 255,
    position.legend = "none",
    fontsize.title = 22, fontsize.labels=c(18,12,8),
    palette= col_palette, ...
  )
}
REVIGO_treemap(stuff)
```
    
To explore the impact of Zika infection DE genes and their functional characteristics, Gene Ontology Term Enrichment Analysis is performed. REVIGO Gene Ontology treemap of DEgene with adjusted p-value < 0.05 and with at least log2 fold changes is shown in _Figure.4_. Main sections of the Gene Ontology treemap are associated with 1: infection( term immune responses, response to other organism); 2: interferon singnaling path way and responses and response to stimulus; which is consistant with privious research and clinical observation of cytokin antivarius responses; 3: negative regulation of multi-organism process, which is interesting and also associated with the speculation that ZIKV host cell showed insufficient antivirus responses.    

```{r}

```



###Reference:    
1. Western Zika Virus in Human Fetal Neural Progenitors Persists Long Term with Partial Cytopathic and Limited Immunogenic Effects
*Natasha W. Hanners, Jennifer L. Eitson, Noriyoshi Usui, et.al.* Cell Rep. 2016 Jun 14; 15(11): 2315–2322.    
2. Brain Metastases from Primary Tumors Epidemiology, Biology, and Therapy 2015, Pages 191-202    
3. AXL-dependent infection of human fetal endothelial cells distinguishes Zika virus from other pathogenic flaviviruses, *Audrey Stéphanie Richard, Byoung-Shik Shim, Young-Chan Kwon*, PNAS February 21, 2017 114 (8) 2024-2029.      


###Appendex:  
  
###Paper origion with dataset:     
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6266559/  
Human Fetal Astrocytes Infected with Zika Virus Exhibit Delayed Apoptosis and Resistance to Interferon: Implications for Persistence. *Daniel Limonta, Juan Jovel, Anil Kumar, et al.* Viruses 2018, 10, 646    
Dataset: 
Next Generation Sequencing Facilitates Quantitative Analysis of A549 Control and A549 infected with Zika virus Transcriptomes (human)
https://www.ncbi.nlm.nih.gov/bioproject/PRJNA610552    

###Record:   
Genome:  
http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.2bit  
Annotation:  
http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/  
hg38.refGene.gtf.gz        10-Jan-2020 09:33   23M  
hg38.trf.bed.gz 

genomeGenerate: --sjdbOverhang 149  
alignReads: --alignIntronMin 20 --alignIntronMax 1000000 --twopassMode Basic  
ENCODE options default for human  
About 5.24% of introns are more than 200,000 bp and less than 10% of introns are more than 11,000 bp in length. Also, < 0.01% of the introns are < 20 bp in length   --M.K. Sakharkar et al. / Distributions of Exons and Introns in the Human Genome  

###Script
Stored in /script useage see README     

###Multi-qc of non-trimed sample fastq files    
![](Graph/Sequence_Count.png){width=400px}, 
![](Graph/Duplication_Levels.png){width=400px}  
![](Graph/Per_Seq_Quality.png){width=400px}
![Sequence Count](Graph/Adapter_Content.png){width=400px}    
  
Samples prepared by poly-A protocol showed higher duplication level, which is expected considering the add-in of adapter sequence. Since all samples passed Fastqc, no further trimming is done for downstream analysis.  

###Alignment QC    

*Figure1. Gene Body Coverage*  
![](Graph/Gene_Body_Cover.png){width=400px}
![](Graph/Reads_Distribution.png){width=400px}  

###Diagnositic plot of Deseq2 fit    
gene-wise estimates congerge towards the fitted estimates, not many outliers is highlighted.  
```{r, echo=F}
plotDispEsts(DESeq.all)
summary(DESeq.all.result)
```
```{r}
sessionInfo()
```

