---
title: "PJ_ANGSD"
author: "JingjingQi"
date: "4/3/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(data.table)
library(DESeq2)
#BiocManager::install("org.Hs.eg.db")
library("TxDb.Hsapiens.UCSC.hg38.knownGene")
library(biomaRt)
library(goseq)
library(org.Hs.eg.db)
```

```{css, echo=FALSE}
.scroll {
  max-height: 200px;
  overflow-y: auto;
}
```

### Introduction  

Source paper: Human Fetal Astrocytes Infected with Zika Virus Exhibit Delayed Apoptosis and Resistance to Interferon: Implications for Persistence. *Daniel Limonta, Juan Jovel, Anil Kumar, et al.* Viruses 2018, 10, 646   

Zika virus (ZIKV) is an arthropod-borne virus(transmited by Arthropod) of the genus Flavivirus(with + sense ssRNA,  10–11kb in length). The recent ZIKV pandemic in theWestern Hemisphere was associated with a dramatic increase in microcephaly and other neurological deficits in the fetuses. Viral kinetics study shown that in general, the half life of infected host cell is less than 10 hours.[1] Previous in-vitro study on human fetal neural progenitors (hNPs) had shown that ZIKV replication persists in hNPs for at least 28 days[2].  hNPs can be differentiated into neurons, oligodendrocytes, or astrocytes. Human fetal astrocytes(HFAs) is the most abundant cell type in the central nervous system, they may function widely as composing the BBB, provision of nutrients to the neurons, maintenance of extracellular ion balance, and damage repair.[3] The researcher of this paper speculated that HFAs possibly functioned as viral reservoirs (cell types or anatomical sites which shown more stable kinetic properties in persistence with replication-competent forms of the virus  than the main pool of actively replicating virus), that may contribute to the establishment of chronic brain infection and neurodevelopmental abnormalities.    

###Logistic:  
Point to prove: virus can enter the host cell, virus keep replicating maintain virulent strength; host cell relatively sustained the virus stress.  
HFAs is permissive to ZIKV infection.(TIM/TAM receptor titer/blockage--ZIKV pcr)    
Low level of apoptosis in HFAs (flow cytometry, A549 control); Virus can thrive in the presence of a sustained cytokine antiviral response(IFN-treatment, virus titer with rtPCR).    
Adaptive mutations were not required for establishing chronic infection. (rtPCR virus genome)   
RNAseq analyses of persistently infected cells revealed that ZIKV alters host gene expression in a manner that could affect developmental processes.   
_Cite from paper:_ [Persistent ZIKV Infection Has A Dramatic Effect on HFA Transcription, the average increase in expression level was 49-fold. Most of the genes that were deregulated by viral infection (|fold change| > 2; FDR < 0.05) were upregulated (Figure 5A,B). In total, 722 transcripts were upregulated, while 81 transcripts were downregulated (Table S2) by chronic ZIKV infection. In persistently infected cells, extensive deregulation of genes that affect morphogenesis of epithelium, cell-substrate adherens junction assembly and focal adhesion assembly were observed (Table S2). The clueGO analysis of upregulated genes identified 10 top terms exclusively related with antiviral defense for persistently infected cells. _Paper Conclusion:_ ZIKV in the fetal brain and that moderate apoptosis combined with inefficient antiviral response from these cells may contribute to the establishment of chronic brain infection associated with the ZIKV neurodevelopmental abnormalities. *control A549 RNA Seq analysis was not reported in the paper]    

### Method   
###Samples 
HFAs were obtained from 15 to 19-week aborted fetuses with written consent.    
Positive Control: continuous human cell lines A549 cells overexpress AXL with high permissive property.    
(HFAs were highly permissive to ZIKV infection, it is shown to be TIM/TAM receptor member AXL dependent.[4], A549 overexpress AXL, depending upon the donor, only 23–46% of HFAs exhibited detectable levels of this protein)   

###RNA Sequencing  
RNA of A549 cells-infected with Zika virus and un-infected A549 cells was isolated using RNeasy mini kit (QIAGEN, Valencia, CA, USA).     
RNAseq libraries were constructed using a TruSeq RNA Sample Prep V2 Kit.     
mRNA profiles was generated by deep sequencing, in triplicate, using Illumina Hiseq4000. All samples passed QC without triming.    

###Project goal:    
Perform bioinformatic analysis to compare with the result with the origional paper that in HFA samples 1: deregulation of genes that affect morphogenesis of epithelium, cell-substrate adherens junction assembly and focal adhesion assembly; 2: upregulated genes identified related with antiviral defense; 3: to find if theres gene level eveidence that support experiment observation of moderate apoptosis and inefficient antiviral response in HFA compare to control A549.    

###Sequence Alignment and Gene Level Expression Quantification  
Sequence alignment and quantification of gene and exon level expression was carried out using RNA-seq analytical pipeline see _appendex_ and _script folder with README_. Pair end reads were aligned to the human ensemble genome GRCh38.99 using star(2.7.0e), Star alignReads command options: --alignIntronMin 20 --alignIntronMax 1000000 --twopassMode Basic. Subread(1.6.2e) was used to count the reads mapping to individual exons according to hg38.refGene annotations (10-Jan-2020). DESeq2(3.10e) Bioconductor packages, is used to normalize gene readscount and gene differencial expression. DEXSeq(1.32.0) Bioconductor packages, is used to normalize exon readscount and transcriptomic differencial expression.    

### Result  
_Table1._ Summary of average aligned reads count and average aligned reads percentage     
```{r, class.output="scroll", echo=F, warning=F, message=F}
r_exon <- fread("data/featCounts_transcript_all.txt.summary.txt")
#rename <- gsub(".Aligned.sortedByCoord.out.bam$", "", x = colnames(r_exon))
#rename <- gsub("/athena/angsd/scratch/jiq4001/Pj1/Star_align/", "", x = rename)
#colnames(r_exon) <- rename
tname <- outer(c("infected", "ctrl"), c("rep1", "rep2", "rep3"), paste)
tname <- outer(c("A549", "HFA1", "HFA2", "HFA3"), c(tname[1,], tname[2,]), paste)
colnames(r_exon) <- c("Status", tname[1, ], tname[2, ], tname[3, ], tname[4, ])
r_exon %>%
  filter(Status %in% c("Assigned", "Unassigned_MultiMapping",
                       "Unassigned_NoFeatures", "Unassigned_Overlapping_Length",
                       "Unassigned_Ambiguity")) %>%
  gather(-Status, key = "sample", value = "reads") %>%
  group_by(sample) %>%
  mutate(percent = round(reads/sum(reads)*100, 2)) -> r_exon_meata

r_exon_meata %>%
  separate(sample, into = c("condition", "rep"))%>%
  group_by(Status) %>%
  summarise(average_aligned_readcount = mean(reads)%>%round(0),
            average_aligned_percentage = mean(percent)%>%round(2),
            sd_rc = sd(reads)%>%round(0),
            sd_pc = sd(percent)%>%round(2)) %>%
  #mutate(reads_count = paste0(average_aligned_rc, "+/-", sd_rc),
  #          reads_percent = paste0(average_aligned_pc, "+/-", sd_pc))%>%
  dplyr::select(Status, average_aligned_readcount, average_aligned_percentage)%>%
  kableExtra::kable()%>%
  kableExtra::kable_styling()
```

The average aligned reads count and average aligned reads percentage of the 6 samples are summarized in _Table.1_. On average, 67.55% aligned reads are uniquely mapped, 16.34% aligned reads are with multi-mapping. 14.76% aligned reads are unassigned due to ambiguity, and less than 2% aligned reads are with no features or too short overlapping length. Individual reads count and percentage summary per sample is shown in the _Figure.1_.     

```{r, echo=F, warning=F, message=F}
#### bar graph alternatime
r_exon_meata %>%
  ggplot()+
  geom_bar(aes(sample, percent, fill = Status), stat = "identity", position = "stack")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        panel.background = element_rect(fill = "white"),
        axis.line = element_line(colour = "black")) -> p1

r_exon_meata %>%
  ggplot()+
  geom_bar(aes(sample, reads, fill = Status), stat = "identity", position = "stack")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        panel.background = element_rect(fill = "white"),
        axis.line = element_line(colour = "black"))-> p2


ggpubr::ggarrange(p1, p2, common.legend = T, legend = "bottom")
```
  
_Figure1._ Summary of average aligned reads count and average aligned reads percentage      

```{r, class.output="scroll", echo=F, message=F, warning=F}
# read data table
rc0 <- read.table("data/featCounts_transcript_all.txt", header=TRUE, row.names = NULL)
#rc1 <- read.table("temp/featCounts_transcript_all_3.txt", header=TRUE, row.names = NULL)
#rc1 <- cbind(rc0, rc1[, 7:12])
#write.table(rc1, file = "data/featCounts_transcript_all.txt",row.names = F)

# trim extra text from colname
names(rc0) <- gsub("X.athena.angsd.scratch.jiq4001.Pj1.Star_align.", "", names(rc0))
names(rc0) <- gsub(".Aligned.sortedByCoord.out.bam", "", names(rc0))

# generate rowname with GeneId
row.names(rc0) <- make.names(rc0$Geneid, unique = T) 
readcounts <- rc0[ , -c(1:6)]


all <- readcounts
namedf <- c("Status", tname[1, ], tname[2, ], tname[3, ], tname[4, ])
names(all) <- namedf[-1]
all_info <- DataFrame(condition = c(rep(c("ctrl", "infected"), each = 3), rep(rep(c("infected", "ctrl"), each = 3), 3)),
                      sample = rep(c("A549", "HFA1", "HFA2", "HFA3"), each = 6),
                      type = c(rep("ctrl", 6), rep("HFA", 18)),
                      row.names = names(all))


# wrap reads, condation into object

DESeq.all <- DESeqDataSetFromMatrix(countData = all, colData = all_info, design = ~ condition + type +  condition:type)
DESeq.HFA <- DESeq.all[, c(7:24)]
DESeq.a <- DESeq.all[, c(1:6)]
DESeq.HFA@design <- DESeq.a@design <- ~condition

# normalization
DESeq.all <- estimateSizeFactors(DESeq.all)
log.norm.counts <- log2(counts(DESeq.all, normalized=TRUE) + 1)
assay(DESeq.all, "log.norm.counts") <- log.norm.counts


DESeq.HFA <- estimateSizeFactors(DESeq.HFA)
log.norm.counts <- log2(counts(DESeq.HFA, normalized=TRUE) + 1)
assay(DESeq.HFA, "log.norm.counts") <- log.norm.counts

DESeq.a <- estimateSizeFactors(DESeq.a)
log.norm.counts <- log2(counts(DESeq.a, normalized=TRUE) + 1)
assay(DESeq.a, "log.norm.counts") <- log.norm.counts

```

Reads counts are normalized on per gene/sample bases with Deseq2 R package, followed by diffrencial expression analysis using GLM with NB distribution model. To investigate the differencially expression of genes induced by KIKV infection, multiple testing on normalized reads count at 0.05 significent level with each gene using glm with contrast $model = condition(Infected_{vs}Ctrl)$ in the subset of HFA samples. To investigate whether HFA response differently than the Control A549 cell in response to KIKV infection, the global differencially analysis is checked using $model =  sample + condition + condition:sample$ on the full reads counts dataset including HFA and A549.    

```{r, echo=F, message=F, warning=F}
DESeq.all <- DESeq(DESeq.all)
DESeq.HFA <- DESeq(DESeq.HFA)
DESeq.a <- DESeq(DESeq.a)
#resultsNames(DESeq.all)
#resultsNames(DESeq.HFA)
DESeq.all.result <- results(DESeq.all, independentFiltering = TRUE, alpha = 0.05,
                             name = "conditioninfected.typeHFA")

DESeq.all.result_HFA <- results(DESeq.HFA, independentFiltering = TRUE, alpha = 0.05,
                             name = "condition_infected_vs_ctrl")
#sum(row.names(DESeq.all.result_HFA) == "ENSG00000137166")
#DESeq.all.result_HFA[row.names(DESeq.all.result_HFA) == "ENSG00000137166",]
DESeq.all.result_a <- results(DESeq.a, independentFiltering = TRUE, alpha = 0.05,
                             name = "condition_infected_vs_ctrl")
#plotDispEsts(DESeq.all)
#plotDispEsts(DESeq.HFA)
#plotDispEsts(DESeq.a)
```

The HFAs sample, 1942 out of 30156 nonzero expressed genes are differencially expressed(with adjusted p-value < 0.05)in the infected group than control gorup, in which 1105 genes are upregulated. Heatmap of the mean normalized Counts of DEgenes with adjusted p-value < 0.05 are plotted in _Figure.2_.  294 out of 33255 nonzero expressede genes induced by ZIKV infection are differencially expressed(with adjusted p-value < 0.05) in HFA compare to control A549, with 217 genes upregulated. Volcano plot of the Log2 fold change vs -Log10 adjusted P-value of DEgenes with adjusted p-value < 0.05 are plotted in _Figure.3_.               
  
```{r, echo=F, message=F, warning=F, eval=F}
summary(DESeq.all.result)
summary(DESeq.all.result_HFA)
summary(DESeq.all.result_a)

table(DESeq.all.result$padj < 0.05)
table(DESeq.all.result_HFA$padj < 0.05)

hist(DESeq.all.result$padj)
plotMA(DESeq.all.result, alpha = 0.05, main = "Test: p.adj.value < 0.05", ylim = c(-8,8))
```

```{r, echo=F, message=F}
DESeq.FHA.result.sorted <- DESeq.all.result_HFA[order(DESeq.all.result_HFA$padj), ]%>% head(table(DESeq.all.result_HFA$padj < 0.05)[2])
#sum(row.names(DESeq.FHA.result.sorted) == "ENSG00000137166")
DE_gene.HFA <- DESeq.HFA@assays@data$log.norm.counts[rownames(DESeq.FHA.result.sorted),]
#colnames(DE_gene.HFA)
rowname_df <- data.frame(Condition = factor(rep(rep(c("Infected", "Ctrl"), each = 3), 3)), 
                         #BioRep = factor(rep(rep(c("Rep1", "Rep2", "Rep3"), 2), 3)),
                         Donor = factor(rep(c("HFA1", "HFA2", "HFA3"), 6)))
rownames(rowname_df) <- colnames(DE_gene.HFA)
pheatmap::pheatmap(DE_gene.HFA, scale = "row", annotation_col = rowname_df, show_rownames = F)


DESeq.all.result.sorted <- DESeq.all.result[order(DESeq.all.result$padj), ]%>% head(table(DESeq.all.result$padj < 0.05)[2])
```
  
_Figure.2_. Heatmap of mean of Normalized Counts of DEgenes in HFAs with adjusted p-value < 0.05.    
      
```{r, echo=F, message=F}
### Genesymbol table query
ensembl = useMart("ensembl",dataset="hsapiens_gene_ensembl") #,host = "useast.ensembl.org"
#listAttributes(ensembl)
id_merge <- getBM(attributes = c("ensembl_gene_id", "external_gene_name", "description", "ensembl_transcript_id"),
                  filters = "ensembl_gene_id", values = row.names(DESeq.FHA.result.sorted), mart = ensembl)%>%data.frame()

named_res.HFA <- DESeq.FHA.result.sorted%>%
  data.frame()%>%
  mutate(ensembl_gene_id = row.names(DESeq.FHA.result.sorted))%>%
  merge.data.frame(id_merge %>% dplyr::select(-ensembl_transcript_id)%>%unique(), by = "ensembl_gene_id", all.x = T)

#named_res.HFA%>%
#  filter(abs(log2FoldChange)>=2)%>%
#  summarise(Meanfold = mean(abs(log2FoldChange)))

named_res.all<- DESeq.all.result.sorted%>%
  data.frame()%>%
  mutate(ensembl_gene_id = row.names(DESeq.all.result.sorted))%>%
  merge.data.frame(id_merge %>% dplyr::select(-ensembl_transcript_id)%>%unique(), by = "ensembl_gene_id", all.x = T)

```
      

```{r, echo=F, warning=F, message=F}
### plot of DE genes aj-pvalue < 0.05
named_res.HFA%>%
  ggplot()+
  geom_point(aes(log2FoldChange, -log10(padj), 
                 color = ifelse(abs(log2FoldChange) >= 2, ">=2", "<")),
             position = position_jitter(width = 0.1, height = 0.1))+
  geom_vline(xintercept = c(2, -2), linetype="dashed", color = "red", size=0.5)+
  labs(title = "HFA_Invected vs non-Infected", x = "log2 fold change", y = "-log10 adjusted-p-value", color = "Absolute\nLog2FoldChange")+
  theme_classic() ->p3

named_res.all%>%
  ggplot()+
  geom_point(aes(log2FoldChange, -log10(padj), 
                 color = ifelse(abs(log2FoldChange) >= 2, ">=2", "<")),
             position = position_jitter(width = 0.1, height = 0.1))+
  geom_vline(xintercept = c(2, -2), linetype="dashed", color = "red", size=0.5)+
  labs(title = "Infected_HFA vs Infected_A549", x = "log2 fold change", y = "-log10 adjusted-p-value", color = "Absolute\nLog2FoldChange")+
  theme_classic() ->p4

#barplot top tail 10 by log2fold
named_res.HFA%>%
  filter(abs(log2FoldChange) >= 2) %>%
  arrange(desc(log2FoldChange))%>%
  mutate(order = 1:n()) %>%
  filter(order %in% c(1:10, (nrow(.)-9):nrow(.)))%>%
  mutate(order = factor(1:n()))->temp_HFA

temp_HFA%>%
  ggplot()+
  geom_bar(aes(order, log2FoldChange, fill = ifelse(log2FoldChange > 0, "> 0", "< 0")), stat = "identity")+
  scale_x_discrete(labels = temp_HFA$external_gene_name)+ 
  labs(title = "HFA_Invected vs non-Infected", x = "Gene Name", y = "Log2 fold change", fill = "Log2FoldChange")+
  coord_flip()+
  theme_classic()->p5

named_res.all%>%
  filter(abs(log2FoldChange) >= 2) %>%
  arrange(desc(log2FoldChange))-> fornagroup

fornagroup%>% 
  filter(!is.na(external_gene_name))%>%
  mutate(order = 1:n())%>%
  filter(order %in% c(1:10, (nrow(.)-9):nrow(.)))%>%
  mutate(order = factor(1:n()))->temp

temp%>%
  ggplot()+
  geom_bar(aes(order, log2FoldChange, fill = ifelse(log2FoldChange > 0, "> 0", "< 0")), stat = "identity")+
  scale_x_discrete(labels = temp$external_gene_name)+ 
  labs(title = "Infected_HFA vs Infected_A549", x = "Gene Name", y = "Log2 fold change", fill = "Log2FoldChange")+
  coord_flip()+
  theme_classic()->p6

ggpubr::ggarrange(p3, p4, p5, p6, common.legend = F, legend = "bottom")

DESeq.FHA.result.sorted.gt2fc <- DESeq.FHA.result.sorted[which(abs(DESeq.FHA.result.sorted$log2FoldChange) >= 2), ] 
#table(DESeq.FHA.result.sorted.gt2fc$log2FoldChange>0)
#summary(DESeq.FHA.result.sorted.gt2fc$baseMean)
#DESeq.FHA.result.sorted.gt2fc[ DESeq.FHA.result.sorted.gt2fc$baseMean == max(summary(DESeq.FHA.result.sorted.gt2fc$baseMean)),]
```
  
_Figure.3_ Volcano plot DEgenes with adjusted p-value < 0.05 and Barplot of top 10 up/down regulated DEgenes.      

To explore the impact of Zika infection DE genes' functional characteristics, Gene Ontology Term Enrichment Analysis is performed on HFA sample DEgene with adjusted p-value < 0.05 and with at least log2 fold changes. REVIGO Gene Ontology treemap is shown in _Figure.4_ with block size corralated with GO term represent frequency.         

```{r, echo=F, message=F, warning=F, eval=F}
gene.vector <- ifelse(DESeq.all.result_HFA$padj < 0.05 & abs(DESeq.all.result_HFA$log2FoldChange) >= 2, 1, 0)
gene.vector <- ifelse(DESeq.all.result_HFA$padj < 0.05 & DESeq.all.result_HFA$log2FoldChange < -2, 1, 0)
gene.vector <- ifelse(row.names(DESeq.all.result_HFA) %in% temp_HFA$ensembl_gene_id, 1, 0)  # top 20 up/down regulated
#sum(gene.vector, na.rm = T)
names(gene.vector) <- rownames(DESeq.all.result_HFA)
gene.vector <- ifelse(is.na(gene.vector), 0, gene.vector)

#glen <- rc0$Length
glen <- getlength(rownames(DESeq.all.result_HFA), "hg38","ensGene")
#sum(gene.vector)
#length(glen)
#length(gene.vector)

gid <- rownames(rc0)
pwf <- nullp(gene.vector, id = gid, bias.data = glen)
go.wall <- goseq(pwf,"hg38","ensGene", use_genes_without_cat=T)
go.gns <- getgo(rownames(DESeq.all.result_HFA), "hg38", "ensGene")%>%stack() 
idx <- p.adjust(go.wall$over_represented_pvalue,method="BH")<.05 | p.adjust(go.wall$under_represented_pvalue,method="BH")<.05
go.wall <- go.wall[idx,]
#write_csv(go.wall, "go.wall_1.csv")
```
  
  
```{r, echo=F, warning=F, message=F}
source("revigo_r/REVIGO_treemap_deseq_IvsN.r")  
REVIGO_treemap <- function(revigo.data, col_palette = "Paired",
                           title = "", ...){
  stuff <- data.frame(revigo.data)
  names(stuff) <- c("term_ID","description","freqInDbPercent","abslog10pvalue",
                    "uniqueness","dispensability","representative")
  stuff$abslog10pvalue <- as.numeric( as.character(stuff$abslog10pvalue) )
  stuff$freqInDbPercent <- as.numeric( as.character(stuff$freqInDbPercent) )
  stuff$uniqueness <- as.numeric( as.character(stuff$uniqueness) )
  stuff$dispensability <- as.numeric( as.character(stuff$dispensability) )
  # check the treemap command documentation for all possible parameters - 
  # there are a lot more
  treemap::treemap(
    stuff,
    index = c("representative","description"),
    vSize = "freqInDbPercent",
    type = "categorical",
    vColor = "representative",
    title = title,
    inflate.labels = FALSE,      
    lowerbound.cex.labels = 0,   
    bg.labels = 255,
    position.legend = "none",
    fontsize.title = 22, fontsize.labels=c(18,12,8),
    palette= col_palette, ...
  )
}
REVIGO_treemap(stuff)
#revigo_g_res <- read.csv("revigo_r/REVIGO_treemap_gene.csv", header = T, skip = 4)
```
  
_Figure.4_ REVIGO Gene Ontology treemap of DEgene with at least log2 fold changes in HFA.         

The fold changes of commonly detected DEgenes of origional paper (detected at transcriptomic level) are compared with our analysis(on average gene level), with result plotted in _Figure.5_    
```{r, echo=F, warning=F, message=F}
paper_res <- readxl::read_xlsx("paper/Table S2, Proofreading Viruses, Nov16,2018.xlsx", sheet = 1, skip = 1)%>%
  mutate(FoldChange_paper = b)%>%
  filter(pVal < 0.05 & FoldChange >= 2)
#
DESeq.FHA.result.sorted.gt2fc$ens_gene <- row.names(DESeq.FHA.result.sorted.gt2fc)
#is.element(DESeq.FHA.result.sorted.gt2fc$ens_gene, paper_res$ens_gene)%>%sum
DESeq.FHA.result.sorted.gt2fc%>%
  data.frame()%>%
  dplyr::select(ens_gene, log2FoldChange) %>%
  inner_join(paper_res %>% dplyr::select(FoldChange_paper, ens_gene, ext_gene))%>%
  group_by(ens_gene)%>%summarise_all(mean)->com_df_gp
Cor_1 = cor(com_df_gp$log2FoldChange, com_df_gp$FoldChange_paper) %>% round(3)
Cor_label_1 = paste0("cor = ", Cor_1)

my_plot <- function(data, Cor_label = "", xt = "", yt="", tt = ""){
  data %>% 
  ggplot(aes(log2FoldChange, FoldChange_paper))+ 
  geom_point()+
  #geom_text(aes(label = target_ID), color = "red", alpha = 0.7)+
  labs(x = xt, y = yt, title = tt)+
  geom_smooth(method = "lm", formula = y ~ x, se = F)+
  geom_hline(yintercept = 0, color = "black")+
  geom_vline(xintercept = 0, color = "black")+
  annotate("text", x = 4, y = 7, label = Cor_label, size = 5)+
  theme(panel.border = element_blank(),
        panel.background = element_blank())
}

p7 <- my_plot(com_df_gp, Cor_label = Cor_label_1, xt = "DESeq", yt="Paper result(Kallisto)", tt = "Fold Change of Commonly Detected DEgenes in Star-subread vs Kallisto")
p7
```

_Figure.5_ Comparison of fold change of commonly detected DEgenes in Star-subread vs Kallisto.    

Analysis of differeicial use of exon is performed using DEXSeq(STAR-HTSeq workflow). Genen groups with differencially expressed exons(adjusted p-value < 0.05) with at least log2-fold change on average are summarized for GO analysis.          
   
Main and interactive effect of cell type vs infection condition is also checked. Differencially expressed genes(adjusted p-value < 0.05) in HFA compared to A549 in presistent ZIKV infected stage with at least log2 fold change are summarized for GO analysis.          

```{r, echo=F, message=F, warning=F, eval=F}
DESeq.all.result.sorted <- DESeq.all.result[order(DESeq.all.result$padj), ]%>% head(table(DESeq.all.result$padj < 0.05)[2])

DE_gene <- DESeq.all@assays@data$log.norm.counts[rownames(DESeq.all.result.sorted),]
rowname_df <- data.frame(Condition = factor(c(rep(c("ctrl", "infected"), each = 3), rep(rep(c("infected", "ctrl"), each = 3), 3))), 
                         type = factor(c(rep("ctrl", 6), rep("HFA", 18))))

rownames(rowname_df) <- colnames(DE_gene)

named_res.all <- DESeq.all.result.sorted%>%
  data.frame()%>%
  mutate(ensembl_gene_id = row.names(DESeq.all.result.sorted))%>%
  merge.data.frame(id_merge, by = "ensembl_gene_id", all.x = T)
named_res.all%>%
  filter(abs(log2FoldChange) >= 2) %>%
  arrange(desc(log2FoldChange))%>%
  mutate(order = 1:n()) %>%
#  filter(order %in% c(1:10, nrow(.)))%>%
mutate(order = factor(1:n()))->temp
table(temp$log2FoldChange>0)
#pheatmap::pheatmap(DE_gene, scale = "row", annotation_col = rowname_df, show_rownames = F)
gene.vector <- ifelse(DESeq.all.result$padj < 0.05 & abs(DESeq.all.result$log2FoldChange) >= 2, 1, 0)
#gene.vector <- ifelse(DESeq.all.result$padj < 0.05 & DESeq.all.result$log2FoldChange < -2, 1, 0)
#sum(gene.vector, na.rm = T)
names(gene.vector) <- rownames(DESeq.all.result)
gene.vector <- ifelse(is.na(gene.vector), 0, gene.vector)

#glen <- rc0$Length
glen <- getlength(rownames(DESeq.all.result), "hg38","ensGene")
#sum(gene.vector)
#length(glen)
#length(gene.vector)
gid <- rownames(rc0)
pwf <- nullp(gene.vector, id = gid, bias.data = glen)
go.wall <- goseq(pwf,"hg38","ensGene", use_genes_without_cat=T)
#go.gns <- getgo(rownames(DESeq.all.result), "hg38", "geneSymbol")%>%stack()
idx <- p.adjust(go.wall$over_represented_pvalue,method="BH")<.05 | p.adjust(go.wall$under_represented_pvalue,method="BH")<.05
go.wall <- go.wall[idx,]
#write_csv(go.wall, "go.wall_ic.csv")
```

```{r, echo=F,message=F, warning=F, eval=F}
### REVIGO Gene Ontology treemap of DEgene with at least log2 fold changes of infection on HFA vs A549. 
source("revigo_r/REVIGO_treemap_deseq_HvsA.r")  
REVIGO_treemap(stuff)

```

### Disscussion:    

Pseudo-aligned with Kallisto default parameters to GRCh38.81 on transcriptomic level was done in the origional paper. 722 transcripts were reported upregulated, while 81 transcripts were reported downregulated on average the of 3 donors, with ajdust-pvalue < 0.05 and greater than log2 fold change. In our analysis, alignment was performed with STAR, followed by subread. At "gene level", 349 defferencially expressed genes showed greater than log2 fold positive change and, 140 DEgenes showed greater than log2 fold negative change with adjust-pvalue < 0.05. Total DEgenes count is inflated with Kallisto aligner compared to STAR as expected, since pseudo aligner failed to properly account for the position information of read fragments. The reported average log2 fold change of DEGs is 8.9 in the paper, our analysis gives 3.5. As to gene level inference, though STAR workflow might lack the power for detecting changes at transcript level, it gives more accurate inference of changes that affect the overall transcriptional output of the gene. Different workflow seems to give uniquely subset of result of differencially expressed genes or transcripts. 134 genes are commonly detected (309 in our analysis, 803 in paper's result), the correlation of commonly detected genes' log2 fold change is 0.926.       

OAS2 (encoded protein induced by interferons, results in viral RNA degradation and the inhibition of viral replication.) was mentioned in the paper as the most upregulated gene in response to ZIKV infection with more than 10 log2-fold change. In our analysis on gene level, OAS2 is also show as one of the top up regulated genes with more than 9 log2-fold change. RSAD2, CMPK2, MX1, MX2, and OASL(know for association to viral infection and immune response to IFN) were also consistantly shown as top upregulated genes as in the origional paper's result. However, FOXP4 gene (known involved in development of the central nervous system) was claimed to be mostly supressed according to the origional paper wasnt shown as a significent result in our analysis.  Down regulated genes in our analysis includes TECTB, which has know association to Alzheimer Disease[7], and SFRP2, which is reported regulate midbrain dopamine neuron development[8].     

Gene Ontology analysis of DEgenes in HFA after presistent KIKV infection show that the top regulated genes are mostly associated with defence response to virus, response to cytokine, regulation of ribonuclease activity (*revigo_r/GO_result_DESeq2.xlsx*), the finding is consistent with the origional paper's analysis. Compare to the origional paper's observation of deregulation of genes that affect morphogenesis of epithelium, cell-substrate adherens junction assembly and focal adhesion assembly, our analysis only found that DEgens associated with response to external stimulus, and biological adhesion.     

In comparing HFA vs A549 control in the infected condition, HFA showed higher level of gene expression that response to cytokine, and protein metabolic process, positive regulation to cell proliferation and cell death. Our analysis did not find stright forward answer to support HFA showed morderate apoptosis or insufficent antivral responses at gene level, since a lot of the viral-host interactions are in balance or each other.(high frequency of positive regulation to cell proliferation was observed in one of the HFA samples) Since moderate apoptosis was confirmed in in-vitro observation, gene level eveidence might require further research or possiblely require lareger sample size.      

Analysis of differeicial use of exon performed using DEXSeq(Star-py-htseq workflow). Associated genes with differencial use of exon are found to be associated with anatomical structure morphogenesis, neuron system development, regulation of developmental process, biological adhesion, which are more consist with origional paper's result of transcript level differencial expression. Interestingly, in comparing GO function associated genes of differeicially use of exon expression between infected HFA vs infected A549, we found high presence of genes that functionally associated with MAPK pathway, which plays important role in reglation of cell proliferation and cell death.      

Overall, our RNAseq analysis did support arguments that HFA shown upregulation of genes associated with cytokin production; differencial use of exons in genes associated with neuron mophology development. No gene level evidence support of inefficient antiviral response in presistently infected HFA. HFA shown differencial use of exons in genes associated with regulation of cell proliferation/death, wich warrent the need of further research.      



###Reference:    
1. Zika plasma viral dynamics in nonhuman primates provides insights into early infection and antiviral strategies *Katharine Best, Jeremie Guedj, Vincent Madelain*, PNAS, 15, 2017 114 (33) 8847-8852.    
2. Western Zika Virus in Human Fetal Neural Progenitors Persists Long Term with Partial Cytopathic and Limited Immunogenic Effects
*Natasha W. Hanners, Jennifer L. Eitson, Noriyoshi Usui, et.al.* Cell Rep. 2016, 14; 15(11): 2315–2322.    
3. Brain Metastases from Primary Tumors Epidemiology, Biology, and Therapy. 2015,191-202    
4. AXL-dependent infection of human fetal endothelial cells distinguishes Zika virus from other pathogenic flaviviruses, *Audrey Stéphanie Richard, Byoung-Shik Shim, Young-Chan Kwon*, PNAS, 21, 2017 114 (8) 2024-2029.      
5. Long Non-Coding RNAs: Emerging and Versatile Regulators in Host–Virus Interactions, *Xing-Yu Meng,1 Yuzi Luo,1 Muhammad Naveed Anwar*,Front Immunol. 2017; 8: 1663.  
6. Differential analyses for RNA-seq: transcript-level estimates improve gene-level inferences
*Charlotte Soneson, Michael I. Love, and Mark D. Robinson*, F1000Res. 2015; 4: 1521.           
7. The striatal kinase DCLK3 produces neuroprotection against mutant huntingtin. *Galvan L, Francelle L, Gaillard MC*, Brain. 2018 May 1;141(5):1434-1454   
8. SFRP1 and SFRP2 Dose‐Dependently Regulate Midbrain Dopamine Neuron Development In Vivo and in Embryonic Stem Cells,
*Julianna Kele  Emma R. Andersson  J. Carlos Villaescusa*, Stem Cells Journals, 30, 5, 2016, 865-875      


###Appendex:  
  
###Paper origion with dataset:     
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6266559/  
Human Fetal Astrocytes Infected with Zika Virus Exhibit Delayed Apoptosis and Resistance to Interferon: Implications for Persistence. *Daniel Limonta, Juan Jovel, Anil Kumar, et al.* Viruses 2018, 10, 646    
Dataset: 
Next Generation Sequencing Facilitates Quantitative Analysis of A549 Control and A549 infected with Zika virus Transcriptomes (human)
https://www.ncbi.nlm.nih.gov/bioproject/PRJNA610552    

###Record:   
Genome:  
Ensemble GRCh38.99.gtf

genomeGenerate: --sjdbOverhang 149  
alignReads: --alignIntronMin 20 --alignIntronMax 1000000 --twopassMode Basic  
ENCODE options default for human  
About 5.24% of introns are more than 200,000 bp and less than 10% of introns are more than 11,000 bp in length. Also, < 0.01% of the introns are < 20 bp in length   --M.K. Sakharkar et al. / Distributions of Exons and Introns in the Human Genome  

###Script
Stored in [script useage see README](https://github.com/jiq4001/NGSA_HW_2020/tree/master/Pj1_Zika_Infection_DEG/script)     

###Multi-qc of non-trimed sample fastq files    
![](Graph/Sequence_Count.png){width=400px}, 
![](Graph/Duplication_Levels.png){width=400px}  
![](Graph/Per_Seq_Quality.png){width=400px}
![Sequence Count](Graph/Adapter_Content.png){width=400px}    
  
A549 ctrl is sequenced at 150bp, HFA is sequenced at 76bp, possibly aim to increase sequencing depth of sample to improve power, though the study only included 3 donor samples.    

###Alignment QC    

*Figure1. Gene Body Coverage*  
![](Graph/Gene_Body_Cover.png){width=400px}
![](Graph/Reads_Distribution.png){width=400px}  

###Diagnositic plot of model fit    
gene-wise estimates congerge towards the fitted estimates, not many outliers is highlighted.   
```{r, echo=F}
cat("DESeq-HFA/A549 ~condition+celltype+condition:celltype")
summary(DESeq.all.result)
```
```{r, echo=F}
cat("DESeq-HFA ~condition")
summary(DESeq.all.result_HFA)
```
```{r, echo=F}
par(mfrow = c(1, 2))
plotDispEsts(DESeq.all, main = "DESeq-HFA/A549 ~condition+celltype+condition:celltype")
plotDispEsts(DESeq.HFA, main = "DESeq-HFA ~condition")
#plotDispEsts( dxd , main = "DEXSeq-HFA ~condition")
```


```{r, message=F, warning=F, echo=F, eval=F}
library(DEXSeq)
countFiles = list.files("DEXSeq_TXT", pattern="*.txt$", full.names=TRUE)
flattenedFile = list.files("DEXSeq_TXT", pattern="gff$", full.names=TRUE)
sampleTable = data.frame(
   row.names = namedf[8:25],
   condition = rep(rep(c("Infected", "Ctrl"), each = 3), 3),
   Donor = rep(c("HFA1", "HFA2", "HFA3"), 6))

dxd = DEXSeqDataSetFromHTSeq(
   countFiles,
   sampleData=sampleTable,
   design= ~sample + celltype + exon + + celltype:exon + condition:exon,
   flattenedfile=flattenedFile )
#exon (‘this’) as well as the count data from the sum of the other exons belonging to the same gene (‘others’)
dxd = estimateSizeFactors( dxd )
dxd = estimateDispersions( dxd) 
dxd = testForDEU( dxd )
dxd1 = estimateExonFoldChanges(dxd, fitExpToVar="celltype")
dxr1 = DEXSeqResults( dxd1 )
rm_na <- dxr1[!is.na(dxr1$padj)& (dxr1$padj < 0.05), ]
dxr1_df <- rm_na[(!is.na(rm_na$log2fold_HFA_A549) & abs(rm_na$log2fold_HFA_A549) >= 2) , ]%>% data.frame()
```


```{r, echo=F, message=F, warning=F, eval=F}
#HFA sample infect vs ctrl
countFiles2 = countFiles[7:24]
sampleTable = data.frame(
     row.names = namedf[-1][7:24],
      condition = c(rep(rep(c("Infected", "Ctrl"), each = 3), 3)),
      celltype = c(rep("HFA", 18)))
dxd = DEXSeqDataSetFromHTSeq(
   countFiles,
   sampleData=sampleTable,
   design= ~sample + celltype + exon + + celltype:exon + condition:exon,
   flattenedfile=flattenedFile )
#exon (‘this’) as well as the count data from the sum of the other exons belonging to the same gene (‘others’)
dxd = estimateSizeFactors( dxd )
dxd = estimateDispersions( dxd) 
dxd = testForDEU( dxd )
#default exons per gene mostly under 40 for human
#>than maxexon or < mincount will cause NA
dxd2 = estimateExonFoldChanges(dxd, fitExpToVar="condition")
dxr2 = DEXSeqResults( dxd2 )
rm_na <- dxr2[!is.na(dxr2$padj)& (dxr2$padj < 0.05), ]
dxr2_df <- rm_na[(!is.na(rm_na$log2fold_Infected_Ctrl) & abs(rm_na$log2fold_Infected_Ctrl) >= 2) , ]%>% data.frame()
#write_csv(dxr2_df, "dexseq_result_HFA_Inf_Ctr.csv")
```


```{r, echo=F, message=F, warning=F, eval=F}
# HFA vs A549 in infected
df <- read.csv("DEXSeq_TXT/dexseq_result_top_HFA_A549.csv")
df <- df[(!is.na(df$log2fold_HFA_A549) & abs(df$log2fold_HFA_A549) >= 2) , ]%>% data.frame() %>%
  arrange(desc(log2fold_HFA_A549))%>%
  mutate(order = 1:n()) %>%
  filter(order %in% c(1:500, (nrow(.)-499):nrow(.)))%>%
  mutate(order = factor(1:n()))


df <- read.csv("DEXSeq_TXT/dexseq_result_HFA_Inf_Ctr.csv")
df <- df[(!is.na(df$log2fold_Infected_Ctrl) & abs(df$log2fold_Infected_Ctrl) >= 2) , ]%>% data.frame() %>%
  arrange(desc(log2fold_Infected_Ctrl))%>%
  mutate(order = 1:n()) %>%
  filter(order %in% c(1:500, (nrow(.)-499):nrow(.)))%>%
  mutate(order = factor(1:n()))

plotDEXSeq( dxr1, "ENSG00000000005", expression=FALSE, splicing=TRUE,   legend=TRUE, cex.axis=1.2, cex=1.3, lwd=2 )
#lapply(unique(query_degdf$groupID), function(x) plotDEXSeq( dxr1, x, expression=FALSE, splicing=TRUE,   legend=TRUE, cex.axis=1.2, cex=1.3, lwd=2 ))
data.frame(exonID = lapply(df$transcripts, function(x) paste(x, collapse = ",")))%>% transpose()%>%
  separate(V1, into = c("e1", "e2", "e3", "e4", "e5", "e6", "e7", "e8", "e9", "e10", 
                        "e11", "e12", "e13", "e14", "e15", "e16", "e17", "e18", "e19")) %>%
  cbind.data.frame(df)%>%
  separate(groupID, into = c("g1", "g2", "g3", "g4", "g5", "g6", "g7", "g8", "g9"))->expand_eg

expand_eg%>%dplyr::select(g1:g9)%>%gather()%>%dplyr::select(value) %>%drop_na()%>%unique()->query_gene
expand_eg%>%dplyr::select(e1:e19)%>%gather()%>%dplyr::select(value) %>%drop_na()%>%unique()->query_exon
query_exon_gene <- getBM(attributes = c("ensembl_gene_id", "external_gene_name", "description", "ensembl_transcript_id"),
                  filters = "ensembl_transcript_id", values = query_exon, mart = ensembl)%>%data.frame()
query_exon_gene <- unique(query_exon_gene$ensembl_gene_id)

gene.vector <- is.element(row.names(DESeq.all.result), unlist(query_gene))
sum(gene.vector)
names(gene.vector) <- rownames(DESeq.all.result)
gene.vector <- ifelse(is.na(gene.vector), 0, gene.vector)
glen <- getlength(rownames(DESeq.all.result), "hg38","ensGene")
gid <- rownames(rc0)
pwf <- nullp(gene.vector, id = gid, bias.data = glen)
go.wall <- goseq(pwf,"hg38","ensGene", use_genes_without_cat=T)
#go.gns <- getgo(rownames(DESeq.all.result), "hg38", "geneSymbol")%>%stack()
idx <- p.adjust(go.wall$over_represented_pvalue,method="BH")<.05 | p.adjust(go.wall$under_represented_pvalue,method="BH")<.05
go.wall <- go.wall[idx,]
#write_csv(go.wall, "go.wall.csv")
source("revigo_r/REVIGO_treemap_dex_HvsA.r")  
source("revigo_r/REVIGO_treemap_dex_IvsN.r") 
REVIGO_treemap(stuff)
```
![Dexseq_HFA_InfectedvsNorm](Graph/Dexseq_HFA_InfectedvsNorm.png){width=400px}
![Dexseq_Infected_HFAvsA549](Graph/Dexseq_Infected_HFAvsA549.png){width=400px}
  
```{r, echo=F, message=F, warning=F, eval=F}
paper_res <- readxl::read_xlsx("paper/Table S2, Proofreading Viruses, Nov16,2018.xlsx", sheet = 1, skip = 1)%>%
  mutate(FoldChange_paper = b)%>%
  filter(pVal < 0.05 & abs(FoldChange) >= 2)

expand_eg%>% dplyr::select(g1:g9, log2fold_Infected_Ctrl)%>%gather(-log2fold_Infected_Ctrl, key = "gene", value = "ens_gene")%>%drop_na()%>%group_by(ens_gene)%>%summarise(log2FoldChange = mean(log2fold_Infected_Ctrl))%>%
  inner_join(paper_res %>% dplyr::select(FoldChange_paper, ens_gene))%>%
  group_by(ens_gene)%>%summarise_all(mean)->com_df_gp1
Cor_2 = cor(com_df_gp1$log2FoldChange, com_df_gp1$FoldChange_paper) %>% round(3)
Cor_label_2 = paste0("cor = ", Cor_2)
#length(unique(com_df_gp1$ens_gene))
expand_eg%>% dplyr::select(e1:e19, log2fold_Infected_Ctrl)%>%gather(-log2fold_Infected_Ctrl, key = "exon", value = "target_ID")%>%drop_na()%>%group_by(target_ID)%>%summarise(log2FoldChange = mean(log2fold_Infected_Ctrl))%>%
  inner_join(paper_res %>% dplyr::select(FoldChange_paper, target_ID))->com_df_ep
Cor_3 = cor(com_df_ep$log2FoldChange, com_df_ep$FoldChange_paper) %>% round(3)
Cor_label_3 = paste0("cor = ", Cor_3)
#length(unique(com_df_ep$target_ID))
plot(com_df_ep$log2FoldChange, com_df_ep$FoldChange_paper)
p8 <- my_plot(com_df_gp1, Cor_label = Cor_label_2, xt = "DEXseq", yt="Paper result(Kallisto)", tt = "Commonly detected DEgens fold change")
p9 <- my_plot(com_df_ep, Cor_label = Cor_label_3, xt = "DEXseq", yt="Paper result(Kallisto)", tt = "Commonly detected DEtranscript fold change")
ggpubr::ggarrange(p7, p8, p9, common.legend = F, legend = "bottom", nrow = 1)
```


```{r}
sessionInfo()
```

