---
title: "HW6"
author: "JingjingQi"
date: "2/12/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##Exercises (14 pts)  
###1 Write a script that will: (3pt)  
    run BWA on one of the samples from the Gierlinski dataset  
    run STAR on the same sample  
Remember those three checks after read alignment:  
Is it a BAM file?   STAR: BAM file by default,  BWA: requires convert  
Is it sorted?  STAR: commond option to sort by coordinate, BWA: requires SAMTOOLS to sort by coordinate  
Is it indexed?  STAR: indexed,   BWA: requires SAMTOOLS to index  
```{eavl = F}
spack load star@2.7.0e
spack load samtools@1.9%gcc@6.3.0
spack load bwa@0.7.15%gcc@6.3.0

wget "http://hgdownload.soe.ucsc.edu/goldenPath/sacCer3/bigZips/sacCer3.2bit"
wget "http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/twoBitToFa"
[jiq4001@node003 sacCer3.fa]$../twoBitToFa sacCer3.2bit sacCer3.fa    ## convert 2bit file to Fasta file


STAR --runMode genomeGenerate \
--runThreadN 4 \
--genomeDir sacCer3_STARindex \
--genomeFastaFiles SacCer3.fa/sacCer3.fasta \   # a file or combined file
--sjdbGTFfile sacCer3.sgd.gtf \
--sjdbOverhang 99

```

```{r, eval=F}
# pooled file from rimed WT biorep1 option -s 10
[jiq4001@node003 trimed_soft]$ for file in $(ls | egrep ".*.gz")   
> do
> zcat $file >> WT_biorep1_cmb
> done

STAR --runMode alignReads \
--runThreadN 4 \
--genomeDir sacCer3_STARindex \
--readFilesIn /athena/angsd/scratch/jiq4001/HW_Wk4/WT/biorep1/trimed_soft/WT_biorep1_cmb \
--outFileNamePrefix Star_align/WT_biorep1_cmb. \
--outSAMattributes NH HI AS nM MD \
--outSAMtype BAM SortedByCoordinate

[jiq4001@node003 Star_align]$ samtools index WT_biorep1_cmb.Aligned.sortedByCoord.out.bam
```
 


```{r, eval=F}
[jiq4001@node003 HW_Wk6]$ bwa index -p sacCer3_BWAindex/sacCer3 SacCer3.fa/sacCer3.fasta # creat index
[jiq4001@node003 HW_Wk6]$ bwa mem -t 4 -k 30 sacCer3_BWAindex/sacCer3  /athena/angsd/scratch/jiq4001/HW_Wk4/WT/biorep1/trimed_soft/WT_biorep1_cmb > BWA_align/WT_biorep1_cmb.bwa.sam 
[jiq4001@node003 HW_Wk6]$ samtools view -b BWA_align/WT_biorep1_cmb.bwa.sam | samtools sort -o BWA_align/WT_biorep1_cmb.bwa.bam
[jiq4001@node003 HW_Wk6]$ rm WT_biorep1_cmb.bwa.sam

```


###2 Subset the aligned reads to select only those that map to chromosome I. (1pt)  
```{r, eval=F}

[jiq4001@node003 Star_align]$ samtools view -h Star_align/WT_biorep1_cmb.Aligned.sortedByCoord.out.bam | egrep "\schrI\s" | wc -l
172283


[jiq4001@node003 HW_Wk6]$ samtools view -h BWA_align/WT_biorep1_cmb.bwa.bam | egrep "\schrI\s" | wc -l
160075

```

###3 Compare the output from BWA and STAR, and summarize any results or differences.  
```{r, eval=F}
STAR
ERR458493.552967  16 chrI 140 255 12M61232N37M2S  *  0  0  CCACTCGTTCACCAGGGCCGGCGGGCTGATCACTTTATCGTGCATCTTGGC     BB?HHJJIGHHJIGIIJJIJGIJIJJIIIGHBJJJJJJHHHHFFDDDA1+B     NH:i:1  HI:i:1  AS:i:41 nM:i:2

BWA
ERR458494.778282  272 chrI  139  0  13H38M  *   0   0       
*   *   NM:i:1  MD:Z:35T2       AS:i:35
```
    ###Which optional SAM fields does STAR add and what do they represent? (1pt)   
    NH: Number of reported alignments that contain the query in the current record  
    HI: Query hit index  
    AS: Alignment score generated by aligner 
    nM: Edit distance to the reference  
    ###Which optional SAM fields does BWA add and what do they represent? (1pt)  
    NM: Edit distance to the reference 
    MD: String for mismatching positions  
    AS: Alignment score generated by aligner  
###4 Run bamqc on your BAM files (Note: this is a tool thatâ€™s not available in spack, but you can use it via /softlib/apps/EL7/BamQC/bin/bamqc after logging on to a compute node). You will need to figure out how to run this on your own (hint: /softlib/apps/EL7/BamQC/bin/bamqc --help).  
```{r, eval=F}
[jiq4001@node003 HW_Wk6]$ /softlib/apps/EL7/BamQC/bin/bamqc -v
[jiq4001@node003 HW_Wk6]$ BamQC v0.1.25_devel

[jiq4001@node003 HW_Wk6]$ /softlib/apps/EL7/BamQC/bin/bamqc Star_align/WT_biorep1_cmb.Aligned.sortedByCoord.out.bam -f sacCer3.sgd.gtf -g SacCer3.fa/ -o Star_align/ --extract

[jiq4001@node003 HW_Wk6]$ /softlib/apps/EL7/BamQC/bin/bamqc BWA_align/WT_biorep1_cmb.bwa.bam -f sacCer3.sgd.gtf -g SacCer3.fa/ -o BWA_align/ --extract

```
    ###Describe 3 differences between the bamqc results for both the BWA and the STAR output files. (3pt)  
    Percent primary alignments: STAR is lower than BWA  
    Percent sequences unmapped: STAR is Lower than BWA  
    Percent sequences spliced: STAR if awear of spice, BWA is not  
###5 Explain the difference between alignment score and mapping quality in SAM/BAM files. How does the interpretation of the mapping quality field differ between STAR and BWA? (2pt)  
Alignment score quantifies the similarity between read and a reference sequence  
Mapping quality quantifies the probability that a read is correctly aligned to the mapped position.  

###6 What is the difference between a multi-mapping read, and a split read? Find a read that has been split in STAR. How did BWA handle the mapping of that read? (2pt)  
multi-mapping read: a sequence that might be aligned to more than one locations in the genome.  
split read: ead that have two or more alignments to the reference from unique region of the read.  
```{r, eval=FALSE}
[jiq4001@node003 HW_Wk6]$ samtools view -h Star_align/WT_biorep1_cmb.Aligned.sortedByCoord.out.bam | head -n 100 | egrep ".*([0-9]+N)+.*"

```
only in Star processed file, BWA doesnt recognize splices.  

###7 How can you remove the unmapped reads from the BWA output? (hint: go back to the notes where FLAG values were explained) (1pt)
```{r, eval=FALSE}
# f output flag, F supress flag    
# unmapped flag code: 4
[jiq4001@node003 HW_Wk6]$ samtools view -b -F 4 BWA_align/WT_biorep1_cmb.bwa.bam > BWA_align/WT_biorep1_cmb.bwa_mapped.bam
```

##Project work (5 pts)  
###If you need a different program than what we have used in the class, you can use spack find to see if the tool is already installed and loadable via spack. If your tool is not there, get in touch with scu@med.cornell.edu to ask them to install it for you.  
If you have processes that will take a long time, go back to the notes from the first day and try to make use of sbatch.  
```{r, eval=FALSE}
spack load -r sra-toolkit@2.8.2-1
```

###1 Download at least one FASTQ file that you will be working with for your project. Document the following details: (2pt)  
    ###where did you get it from?   
    https://www.ncbi.nlm.nih.gov/Traces/study/?acc=SRP026013&o=rna_preparation_sam_s%3Aa%3Bacc_s%3Aa&s=SRR900275,SRR900276,SRR900277,SRR900278,SRR900279,SRR900280,SRR900281,SRR900282,SRR900283,SRR900284,SRR900285,SRR900286,SRR900287,SRR900288,SRR900289,SRR900290  
    ###what publication is it linked to?   
    https://www.ncbi.nlm.nih.gov//pubmed/23977132   
    ###who generated the data?  
    Mayo Clinic College of Medicine  
    ###how was the NA extracted?  
    TRIzol kit (Invitrogen)  
    ###what library prep was used?  
    Illumina PolyA, NuGEN Ovation  
    ###what cell type was used?  
    human mammary epithelial cell (HMEC) lines obtained from American Type Culture Collection (ATCC)  
    ###what was the treatment/experimental condition? 
    involuted phenotype associated with an increased breast cancer risk  
    ###what sequencing platform was used?
    NuGEN---GAIIx sequencer, and Illumina PolyA---HiSeq 2000 at Mayo Clinical  
    
###2 Align the FASTQ file with an appropriate aligner (you may have to build a new index). Document: (3pt)  
    parameters (and why you chose them)  
    summary of outcome and basic QC  
###Resources: 
http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/
hg38.2bit- contains the complete human/hg38 genome sequence
    in the 2bit file format.  Repeats from RepeatMasker and Tandem Repeats
    Finder (with period of 12 or less) are shown in lower case; non-repeating
    sequence is shown in upper case.  
GRCh38
Date: 2013/12/17
Assembly type: haploid-with-alt-loci
Release type: patch
Assembly level: Chromosome 

http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/  
hg38.ensGene.gtf.gz       10-Jan-2020 09:33   27M  

https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-S11-S8 # impack of annotation choice  

```{r, eval=FALSE}
[jiq4001@node003 Pj]$ wget "http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.2bit"
[jiq4001@node003 Pj]$/athena/angsd/scratch/jiq4001/HW_Wk6/twoBitToFa hg38.2bit hg38.fa

[jiq4001@node003 Pj]$ wget "http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/hg38.ensGene.gtf.gz"

#!/bin/bash
#SBATCH --nodes=1
#SBATCH --ntasks=1 --cpus-per-task=16
#SBATCH --job-name=tmle
#SBATCH --time=7-00:00:00
#SBATCH --mem=80G
#SBATCH --error=/athena/angsd/scratch/jiq4001/error

spack load star@2.7.0e

cwd=/athena/angsd/scratch/jiq4001/Pj
  
STAR --runMode genomeGenerate \
--runThreadN 6 \
--genomeDir /athena/angsd/scratch/jiq4001/Pj/hg38_STARindex   \
--genomeFastaFiles /athena/angsd/scratch/jiq4001/Pj/hg38.fa   \
--sjdbGTFfile /athena/angsd/scratch/jiq4001/Pj/hg38.ensGene.gtf  \
--sjdbOverhang 99 #????????? where to find read length
```

```{r, eval=FALSE}
#!/bin/bash 
spack load fastqc

# set current work dir
cwd=/athena/angsd/scratch/jiq4001/Pj/
mkdir ${cwd}/QC
mkdir ${cwd}/trimed_QC
# file acc
creat Acc_ilumina.txt  # Acc_Nugen.txt
creat Adap_ilumina.txt # Adap_Nugen.txt


for file in $(cat Acc_ilumina.txt)
  do
    fastqc $file -o ./QC/$file -t 4 --extract
done


```

